### ISSUE TITLE
isMatch transform option not preventing transforms

### ISSUE THREAD

**Issue by CGastrell** [WordPress/gutenberg#31769]
## Description
Returning `false` from `isMatch` option on a transform is not preventing the transformation to be listed nor the block to be transformed.

On top of that, `isMatch` evaluation should come in sooner, so `transform` function is not required if `isMatch` returns `false`.

## Step-by-step reproduction instructions
On a test/development block, provide the transform metadata to prevent any transformation:

(Note that I'm forced to include a transform function, even if it's never going to happen, this is why the 2 buggy behaviors are related. For the sake of sample code, let's assume the block name is `test/block`)

```js
transforms: {
	to: [
		{
			type: 'block',
			blocks: [ '*' ],
			transform: ( attrs ) => createBlock( 'test/block', attrs, [] ),
			isMatch: () => false,
		},
	],
},
```
Once the block is built and loaded, insert it on a post and click on the "transform to" button on the toolbar.

## Expected behaviour
According to [docs](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-transforms/) there shouldn't be any transformations listed. `isMatch` evaluation should prevent it.

## Actual behaviour
Group and Column are still available as transformations (screenshot is from a Jetpack enabled site, which also shows "Premium Content" as an available transform).

## Screenshots or screen recording (optional)
![image](https://user-images.githubusercontent.com/157240/118017815-6a1b3400-b32d-11eb-845c-668858815050.png)

## Code snippet (optional)
```
transforms: {
	to: [
		{
			type: 'block',
			blocks: [ '*' ],
			transform: ( attrs ) => createBlock( 'test/block', attrs, [] ),
			isMatch: () => false,
		},
	],
},
```

## WordPress information
- WordPress version: 5.7.1
- Gutenberg version: 10.5.4
- Are all plugins except Gutenberg deactivated? No
- Are you using a default theme (e.g. Twenty Twenty-One)? Yes

## Device information
- Device: Desktop
- Operating system: Mac OS 11.2.2
- Browser: Chrome / Firefox / Safari / Edge

**Comment by stokesman** [comment]
As I understand it the `isMatch` is only meant to determine whether its associated `transform` will be available and not to affect availability of other transforms specified by other blocks (In this case the `from` transforms specified by Columns, Group and Premium Content).

I'm unaware of anyway for a block to deny transforms specified by other blocks but seems like it might be a valid use case.

**Comment by CGastrell** [comment]
> As I understand it the `isMatch` is only meant to determine whether its associated `transform` will be available and not to affect availability of other transforms specified by other blocks (In this case the `from` transforms specified by Columns, Group and Premium Content).


That might be the case, but if it is, [documentation](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-transforms/) should be changed then:
![image](https://user-images.githubusercontent.com/157240/118133787-9b503e80-b3d7-11eb-9a71-362cb9b75dae.png)

Beyond anyone's particular understanding, the text is quite clear about it. Now [reading the code](https://github.com/WordPress/gutenberg/blob/6abd194a333beccff54f031fa8b396578d7ae4ff/packages/blocks/src/api/factory.js#L169-L241) it also looks as if the existence of `isMatch` should suffice (returning `false`) to make a transformation *not possible*

> I'm unaware of anyway for a block to deny transforms specified by other blocks but seems like it might be a valid use case.

Then again, if you're right about the actual purpose of `isMatch`, maybe I'm barking at the wrong tree and this should be more on the *feature request* side of things rather than a bug report. Such feature would be *provide an option to opt out of/disallow transforms, even from '\*' ones*. Maybe some:
```js
transforms: {
  to: [
    {
      type: 'block',
      blocks: [ '*' ],
      transform: () => false, // deny any transformation by evaluating the transform function (might be expensive) or
      allowTransform: () => false, // add an actual check to be run before any other checks (what I thought isMatch did)
    }
  ]
}
```
Shall that be the case, this issue should be dismissed and I'll go for a feature request.

**Comment by ntsekouras** [comment]
>this should be more on the feature request side of things rather than a bug report. Such feature would be provide an option to opt out of/disallow transforms, even from '*' ones. Maybe some:

Exactly. The `isMatch` is related to the current transformation you're declaring it and not other possible transforms. This functionality doesn't exist and is a combination of a flag like you're referring to `disallow` and wildcard character `*`.

**Comment by CGastrell** [comment]
Closing in favor of a feature request #31905
